File: algorithms/1_easy/F/03471_find_the_largest_almost_missing_integer/solution.c
#define MAX_NUM 50
#define MIN_NUM 0
#define TABLE_LEN (MAX_NUM - MIN_NUM + 1)

int largestInteger(const int *restrict nums, const int len, const int k) {
    if (len == 0 || nums == NULL || k > len || k < 0) { return -1; }

    int res = -1;

    if (k == len) {
        for (int i = 0; i < len; i += 1) {
            res = MAX(res, nums[i]);
        }
    } else if (k == 1) {
        int freqs[TABLE_LEN] = { 0 };

        for (int i = 0; i < len; i += 1) {
            if (nums[i] < MIN_NUM || nums[i] > MAX_NUM) { return -1; }

            freqs[nums[i] - MIN_NUM] += 1;
        }

        for (int i = TABLE_LEN - 1; i >= 0; i -= 1) {
            if (freqs[i] == 1) {
                res = i + MIN_NUM;
                break;
            }
        }
    } else {
        const int first = nums[0];
        const int last = nums[len - 1];
        int cnt_first = 0;
        int cnt_last = 0;

        for (int i = 0; i < len; i += 1) {
            if (nums[i] == first) {
                cnt_first += 1;
            }

            if (nums[i] == last) {
                cnt_last += 1;
            }
        }

        if (cnt_first == 1 && cnt_last == 1) {
            res = MAX(first, last);
        } else if (cnt_first == 1) {
            res = first;
        } else if (cnt_last == 1) {
            res = last;
        }
    }

    return res;
}

File: algorithms/1_easy/F/03471_find_the_largest_almost_missing_integer/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn largest_integer(nums: Vec<i32>, k: i32) -> i32 {
        assert!(k >= 1 && (k as usize) <= nums.len());
        let len = nums.len();
        let k = k as usize;
        let mut res: Option<i32> = None;

        if k == len {
            res = Some(*nums.iter().max().unwrap());
        } else if k == 1 {
            let mut freqs: HashMap<i32, bool> = HashMap::new();

            for &num in &nums {
                *freqs.entry(num).and_modify(|x| *x = false).or_insert(true);
            }

            for (num, state) in freqs {
                if state {
                    res = match res {
                        None => Some(num),
                        Some(val) => Some(val.max(num)),
                    }
                }
            }
        } else {
            let (first, last) = (nums[0], nums[len - 1]);
            let (mut cnt_first, mut cnt_last) = (0, 0);

            for &num in &nums {
                if num == first {
                    cnt_first += 1;
                }

                if num == last {
                    cnt_last += 1;
                }
            }

            if cnt_first == 1 && cnt_last == 1 {
                res = Some(first.max(last));
            } else if cnt_first == 1 {
                res = Some(first);
            } else if cnt_last == 1 {
                res = Some(last);
            }
        }

        match res {
            None => -1,
            Some(val) => val,
        }
    }
}
